lib/sigstore/signer.rb

```diff
@@ -0,0 +1,246 @@
+
+    def verify_chain(chain)
+      # Perform certification path validation (RFC 5280 §6) of the returned certificate chain with the pre-distributed
+      # Fulcio root certificate(s) as a trust anchor.
+
+      x509_store = OpenSSL::X509::Store.new
+      chain = chain.map(&:openssl)
+      leaf = chain.shift
+      root = chain.pop # TODO: use root from trusted root as trust anchor
+      x509_store.add_cert root
+      raise x509_store.error_string unless x509_store.verify(leaf, chain)
+
+      logger.debug { "verified chain" }
+
+      # TODO: verify SCT in leaf
+      # Extract a SignedCertificateTimestamp, which may be embedded as an X.509 extension in the leaf certificate or
+      # attached separately in the SigningCertificate returned from the Identity Service.
+      # Verify this SignedCertificateTimestamp as in RFC 9162 §8.1.3, using the root certificate from
+      # the Certificate Transparency Log.
+
+      # TODO: verify leaf subject
+      # Check that the leaf certificate contains the subject from the certificate signing request and encodes the
+      # appropriate AuthenticationServiceIdentifier in an extension with OID 1.3.6.1.4.1.57264.1.8.
+
+      [leaf, x509_store.chain]
+    end
+
+    def sign_payload(payload, key)
+      # TODO: derive correct digest from what the registry supports
+      # The Signer MAY pre-hash the payload using a hash algorithm from the registry (Spec: Sigstore Registries) for
+      # compatibility with some signing metadata formats (see §Submission of Signing Metadata to Transparency Service).
+      key.sign("SHA256", payload)
+    end
+
+    # TODO: implement
+    def submit_signature_hash_to_timstamping_service(_signature)
+      # The Signer sends a hash of the signature as the messageImprint in a TimeStampReq to the Timestamping Service and
+      # receives a TimeStampResp including a `TimeStampToken`.
+      # The signer MUST verify the TimeStampToken against the payload and Timestamping Service root certificate.
+
+      {}
+    end
+
+    def submit_signed_metadata_to_transparency_service(signature, cert, data_sha256)
+      # The Signer chooses a format for signing metadata; this format MUST be in the supportedMetadataFormats in the
+      # Transparency Service configuration. The Signer prepares signing metadata containing at a minimum:
+      # * The signature.
+      # * The payload (possibly pre-hashed; if so, the entry also includes the identifier of the hash algorithm).
+      # * Verification material (signing certificate or verification key).
+      #   * If the verification material is a certificate, the client SHOULD upload only the signing certificate and
+      #     SHOULD NOT upload the CA certificate chain.
+      #
+      # The signing metadata might contain additional, application-specific metadata according to the format used.
+      # The Signer then canonically encodes the metadata (according to the chosen format).
+
+      # TODO: is looping here correct?
```

Comment:
can there be multiple tlogs to upload to?

Reply:
Yes, although the client *should* probably just pick one to upload to for now. I don't think there's any harm in uploading to all of them, but most clients expect only a single tlog entry in bundles for now.